tdrs-backend/tdpservice/search_indexes/parsers/preparser.py

```diff
@@ -0,0 +1,210 @@
+    elif len(line) != 23:
+        logger.debug("line: '%s' len: %d", line, len(line))
+        return False, {'preparsing': 'Header length incorrect.'}
+
+    return True, line
+
+def get_trailer_line(datafile):
+    """Alters the trailer line into usable string."""
+    # certify/transform input line to be correct form/type
+
+    # Don't want to read whole file, just last line, only possible with binary
+    # Credit: https://openwritings.net/pg/python/python-read-last-line-file
+    # https://stackoverflow.com/questions/46258499/
+    try:
+        datafile.seek(-2, os.SEEK_END)  # Jump to the second last byte.
+        while datafile.read(1) != b'\n':  # Check if new line.
+            datafile.seek(-2, os.SEEK_CUR)   # Jump two bytes back
+    except OSError:  # Either file is empty or contains one line.
+        datafile.seek(0)
+        return False, {'preparsing': 'File too short or missing trailer.'}
+
+    # Having set the file pointer to the last line, read it in.
+    line = datafile.readline().decode()
+    datafile.seek(0)  # Reset file pointer to beginning of file.
+
+    logger.info("Trailer line: '%s'", line)
+    if get_record_type(line) != 'TR':
+        raise ValueError('Last line is not recognized as a trailer line.')
+    elif len(line) != 24:
+        logger.debug("line: '%s' len: %d", line, len(line))
+        return False, {'preparsing': 'Trailer length incorrect.'}
+    line = line.strip('\n')
+
+    return True, line
+
+def preparse(data_file, data_type, section):
+    """Validate metadata then dispatches file to appropriate parser."""
+    if isinstance(data_file, DataFile):
+        logger.debug("Beginning preparsing on '%s'", data_file.file.name)
+        datafile = data_file.file  # do I need to open() this?
+    elif isinstance(data_file, BufferedReader):
+        datafile = data_file
+    else:
+        logger.error("Unexpected datafile type %s", type(data_file))
+        raise TypeError("Unexpected datafile type.")
+
+    header_preparsed, line = get_header_line(datafile)
+    if header_preparsed is False:
+        raise ValueError("Header invalid, error: %s" % line['preparsing'])
+    # logger.debug("Header: %s", line)
+
+    header_is_valid, header_validator = validate_header(line, data_type, section)
+    if isinstance(header_validator, Exception):
+        raise header_validator
+
+    trailer_preparsed, line = get_trailer_line(datafile)
+    if trailer_preparsed is False:
+        raise ValueError("Trailer invalid, error: %s" % line['preparsing'])
+    trailer_is_valid, trailer_validator = validate_trailer(line)
+    if isinstance(trailer_validator, Exception):
+        raise trailer_validator
+
+    errors = {'header': header_validator.errors, 'trailer': trailer_validator.errors}
+
+    if header_is_valid and trailer_is_valid:
+        logger.info("Preparsing succeeded.")
+    else:
+        # TODO: should we end here or let parser run to collect more errors?
```

Comment:
@andrew-jameson good q. happy to discuss. i think the parity standard is that nothing changes with how STTs prepare the files. so we expect headers and trailers and we expect them to be correct. i see preparsing failure (e.g. quarter out of range) currently prevents parsing, i presume there's a ticket to inform the user of the preparse failure.

Reply:
But I'll also add that the header details are ultimately not more important than the combo box selections and file type selections the user makes before uploading files.

Reply:
@ADPennington Let me know if there's anything to do here -- my interpretation was that there isn't. If not, please resolve this conversation.